modules/stochastic_tools/include/utils/StochasticToolsUtils.h

```diff
@@ -0,0 +1,210 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MooseUtils.h"
+#include "libmesh/communicator.h"
+
+namespace StochasticTools
+{
+
+/**
+ * Custom type trait that has a ::value of true for types that can be gathered
+ */
+template <typename T>
+struct canDefaultGather
+{
+  static constexpr bool value = false;
+};
+template <typename T>
+struct canDefaultGather<std::vector<T>>
+{
+  static constexpr bool value = std::is_base_of<TIMPI::DataType, TIMPI::StandardType<T>>::value;
+};
+template <typename T>
+struct canStochasticGather
+{
+  static constexpr bool value = false;
+};
+template <typename T>
+struct canStochasticGather<std::vector<T>>
+{
+  static constexpr bool value = canStochasticGather<T>::value ||
+                                std::is_base_of<TIMPI::DataType, TIMPI::StandardType<T>>::value ||
+                                std::is_same<T, std::string>::value || std::is_same<T, bool>::value;
+};
+
+/*
+ * Methods for gathering nested vectors
+ */
+template <typename T>
+void
+stochasticGather(const libMesh::Parallel::Communicator &, processor_id_type, T &)
+{
+  ::mooseError("Cannot gather values of type ", MooseUtils::prettyCppType<T>());
```

Comment:
You should be able to make this a compile time error.

```
static_assert(canStochasticGather<T>::value, "This type is not implemented for static gather");
```

Reply:
The issue with this is in `StochasticReporterContext` where I still want to be able to construct values that aren't able to be gathered, but throw an error if we try. Would you prefer this type of logic in `StochasticReporterContext` instead of here?

Reply:
I think it's fine here.